src/communicate.rs

```diff
@@ -306,7 +306,8 @@ fn handle_echo_resp(our_ext_addr: SocketAddr, inform_tx: Option<Sender<SocketAdd
     }
 }
 
-fn handle_communication_err(peer_addr: SocketAddr, e: &Error, details: &str) {
+/// When connection fails, remove it to prevent memory leaks.
+pub fn handle_communication_err(peer_addr: SocketAddr, e: &Error, details: &str) {
```

Comment:
Doesn't need to be public - see other reviews.
While at it can you pls change:
https://github.com/ustulation/using-quinn/blob/master/src/communicate.rs#L315-L319
to simply remove only (like said for the review in `listener.rs`) as otherwise it's going to cause a spurious fire.
So only `let _ = c.connections.remove(..);`